x/cronos/keeper/precompiles/ica.go

```diff
@@ -0,0 +1,218 @@
+	return &IcaContract{
+		BaseContract:  NewBaseContract(IcaContractAddress),
+		cdc:           cdc,
+		icaauthKeeper: icaauthKeeper,
+	}
+}
+
+func (ic *IcaContract) Address() common.Address {
+	return IcaContractAddress
+}
+
+// RequiredGas calculates the contract gas use
+func (ic *IcaContract) RequiredGas(input []byte) uint64 {
+	return ICAContractRequiredGas
+}
+
+func (ic *IcaContract) IsStateful() bool {
+	return true
+}
+
+func (ic *IcaContract) Run(evm *vm.EVM, contract *vm.Contract, readonly bool) ([]byte, error) {
+	// parse input
+	methodID := contract.Input[:4]
+	stateDB := evm.StateDB.(ExtStateDB)
+	precompileAddr := ic.Address()
+	converter := cronosevents.IcaConvertEvent
+	var execErr error
+	var res codec.ProtoMarshaler
+	switch string(methodID) {
+	case string(RegisterAccountMethod.ID):
+		if readonly {
+			return nil, errors.New("the method is not readonly")
+		}
+		args, err := RegisterAccountMethod.Inputs.Unpack(contract.Input[4:])
+		if err != nil {
+			return nil, errors.New("fail to unpack input arguments")
+		}
+		connectionID := args[0].(string)
+		account := args[1].(common.Address)
+		version := args[2].(string)
+		txSender := evm.Origin
+		if !isSameAddress(account, contract.CallerAddress) && !isSameAddress(account, txSender) {
+			return nil, errors.New("unauthorized account registration")
+		}
+		owner := sdk.AccAddress(account.Bytes()).String()
+		execErr = stateDB.ExecuteNativeAction(precompileAddr, converter, func(ctx sdk.Context) error {
+			response, err := ic.icaauthKeeper.RegisterAccount(ctx, &types.MsgRegisterAccount{
+				Owner:        owner,
+				ConnectionId: connectionID,
+				Version:      version,
+			})
+			res = response
+			return err
+		})
+	case string(QueryAccountMethod.ID):
+		args, err := QueryAccountMethod.Inputs.Unpack(contract.Input[4:])
+		if err != nil {
+			return nil, errors.New("fail to unpack input arguments")
+		}
+		connectionID := args[0].(string)
+		account := args[1].(common.Address)
+		owner := sdk.AccAddress(account.Bytes()).String()
+		execErr = stateDB.ExecuteNativeAction(precompileAddr, nil, func(ctx sdk.Context) error {
+			response, err := ic.icaauthKeeper.InterchainAccountAddress(ctx, &types.QueryInterchainAccountAddressRequest{
+				Owner:        owner,
+				ConnectionId: connectionID,
+			})
+			res = response
+			return err
+		})
+	case string(SubmitMsgsMethod.ID):
+		if readonly {
+			return nil, errors.New("the method is not readonly")
+		}
+		args, err := SubmitMsgsMethod.Inputs.Unpack(contract.Input[4:])
+		if err != nil {
+			return nil, errors.New("fail to unpack input arguments")
+		}
+		connectionID := args[0].(string)
+		account := args[1].(common.Address)
+		data := args[2].(string)
+		timeout := args[3].(*big.Int)
+		txSender := evm.Origin
+		if !isSameAddress(account, contract.CallerAddress) && !isSameAddress(account, txSender) {
+			return nil, errors.New("unauthorized send tx")
+		}
+		var icaMsgData icatypes.InterchainAccountPacketData
```

Comment:
Maybe better to get the msg and construct icatypes.InterchainAccountPacketData from the msg

Reply:
user/contract should provide `data`, precompile should hardcode `type` as `EXECUTE_TX`.

Reply:
fixed